pkg/types/ccipocr3/generic_types.go

```diff
@@ -82,51 +82,85 @@ func (c ChainSelector) String() string {
 	return fmt.Sprintf("ChainSelector(%d)", c)
 }
 
-type CCIPMsg struct {
-	CCIPMsgBaseDetails
-	ChainFeeLimit   BigInt        `json:"chainFeeLimit"`
-	Nonce           uint64        `json:"nonce"`
-	Sender          types.Account `json:"sender"`
-	Receiver        types.Account `json:"receiver"`
-	Strict          bool          `json:"strict"`
-	FeeToken        types.Account `json:"feeToken"`
-	FeeTokenAmount  BigInt        `json:"feeTokenAmount"`
-	Data            []byte        `json:"data"`
-	TokenAmounts    []TokenAmount `json:"tokenAmounts"`
-	SourceTokenData [][]byte      `json:"sourceTokenData"`
-	// Metadata is used as a backup for any additional data that is not covered by the fields above.
-	Metadata CCIPMsgMetadata `json:"metadata"`
-}
-
-type CCIPMsgMetadata struct {
-	// Version of the message metadata. Required in order to be able to parse the metadata
-	// by the underlying implementation.
-	Version string `json:"version"`
-	// Data is the metadata payload. The underlying implementation should know how to parse this data.
-	Data []byte `json:"data"`
-}
-
-type TokenAmount struct {
-	Token  types.Account
-	Amount *big.Int
-}
-
-func (c CCIPMsg) String() string {
+// Message is the generic Any2Any message type for CCIP messages.
+// It represents, in particular, a message emitted by a CCIP onramp.
+// The message is expected to be consumed by a CCIP offramp after
+// translating it into the appropriate format for the destination chain.
+type Message struct {
+	// Header is the family-agnostic header for OnRamp and OffRamp messages.
+	// This is always set on all CCIP messages.
+	Header RampMessageHeader `json:"header"`
+
+	// EVM2AnyRampMessage is populated if the CCIP message originated from an EVM chain.
+	EVM2AnyRampMessage *EVM2AnyRampMessage `json:"evm2AnyRampMessage,omitempty"`
```

Comment:
Maybe include another field e.g. `Origin="evm"` so we can do:
```
switch Origin:
   case "evm":
      use msg.EVM2AnyRampMessage
```

otherwise we have to:
```
if msg.EVM2AnyRampMessage != nil {
  ...
}
```

Reply:
I was imagining there'd be an API in `chain-selectors` that will do this:

```
switch chainsel.GetChainFamily(msg.DestChainSelector) {
  case chainsel.EVMFamily:
    // ...
  case chainsel.SolanaFamily:
    // ...
}
```

Reply:
I have another suggestion.

```
// Base struct
type IMessage interface {
	GetHeaders() RampMessageHeader
}

type struct WEVM2AnyRampMessage {
   message EVM2AnyRampMessage
}

func (msg * WEVM2AnyRampMessage)  GetHeaders() {
 ....
}

func (msg * WEVM2AnyRampMessage)  SomeOtherFunc() {
 ....
}

switch msg.(type) {
case WEVM2AnyRampMessage:
  msg.SomeOtherFunc()
}
```

Should be more resilient than a type string like "evm". 


> I was imagining there'd be an API in chain-selectors that will do this:
```
switch chainsel.GetChainFamily(msg.DestChainSelector) {
  case chainsel.EVMFamily:
    // use EVM2Any
  case chainsel.SolanaFamily:
    // use Sol2Any
}
```
Is this the only place that it will be called from?

Reply:
^ agreed on above, on-chain I'm currently thinking:

* `chainFamilyTag` gets assigned to a dest chain config, e.g. for ETH, ARB, etc. assign `EVM_FAMILY_TAG`
* Each family tag is generated deterministically, e.g. right now I use `bytes4(keccak256("CCIP EVM Family"))` (32-bit family ids should be more than enough I recon)